### ISSUE TITLE
`live preview` doesn't work

### ISSUE THREAD

**Issue by frankfarzan** [GoogleContainerTools/kpt#1929]
Client-side preview doesn't really preview changes that will be applied:

Change a single field in wordpress package (e.g. port value). There's a discrepency between `preview` and `apply` commands:

```
wordpress [main✗] » kpt live preview
service/wordpress configured (preview)
service/wordpress-mysql configured (preview)
deployment.apps/wordpress configured (preview)
deployment.apps/wordpress-mysql configured (preview)
persistentvolumeclaim/mysql-pv-claim configured (preview)
persistentvolumeclaim/wp-pv-claim configured (preview)
6 resource(s) applied. 0 created, 0 unchanged, 6 configured, 0 failed (preview)
0 resource(s) pruned, 0 skipped, 0 failed (preview)
 
wordpress [main✗] » kpt live apply
service/wordpress configured
service/wordpress-mysql unchanged
deployment.apps/wordpress unchanged
deployment.apps/wordpress-mysql unchanged
persistentvolumeclaim/mysql-pv-claim unchanged
persistentvolumeclaim/wp-pv-claim unchanged
6 resource(s) applied. 0 created, 5 unchanged, 1 configured, 0 failed
0 resource(s) pruned, 0 skipped, 0 failed
```

**Comment by mortent** [comment]
This is the expected behavior here. By default kpt will do a client-side dry-run, which doesn't actually compare the state of the local resources against the resources in the cluster. So it doesn't have any way to determine if a resource will be updated or remain unchanged when we do the actual apply. Thus, the output when actually doing the apply (when we do know what happened) will differ.

In general though, `kpt live` is built on top of kubectl libraries and we don't have anything else to work with here other than the information returned. The categories used here is the same as used in kubectl. Working around this would mean either make improvements in upstream kubectl to provide more information (and even here we might be limited to what is coming back from the apiserver), or break some of the dependencies on kubectl code and implement it ourselves. Both would be substantial changes.

**Comment by frankfarzan** [comment]
I would flip the question around. Why should a user run the `preview` command at all? Specifically for the example above, what is it telling us?

Tangential to this is that the question of statistics printed for apply and preview, e.g.:

```
6 resource(s) applied. 0 created, 0 unchanged, 6 configured, 0 failed (preview)
0 resource(s) pruned, 0 skipped, 0 failed (preview)
```

What are we telling the user here in plain english? What's does `6 configured` mean? We should file a separate issue for this. As it stands, it's hard to parse and the semantic of each category is not clear.

**Comment by mortent** [comment]
Preview is useful for two things:
* Validation of the resources prior to applying any of them. Client-side dry-run does some validation, but this is much more useful with server-side dry-run.
* See which resources will be applied and which resources will be pruned. The example here doesn't have an example of that, but being able to see which resources will be pruned is quite useful. As we implement more lifecycle directives, it can be pretty useful to know which resources are impacted.

Now I agree that the information about which operation happens for each resource might be a little bit unclear and it can be confusing that they don't align between `preview` and `apply`. But the current categories are aligned with kubectl and they basically is mapped as follow:
 * created => resource didn't exist already, so it was created.
 * unchanged => resource already exists and it wasn't changed.
 * configured => resource already exists and it has been updated.

Other than `configured`, the categories seems pretty straight-forward.

As kubernetes will soon move completely to server-side apply, my understanding is that these categories will no longer be relevant. Every resource will just be `applied` server-side and it doesn't really matter if it was create, update or no-op. So the output from apply would end up pretty much similar to prune, where we just list how many resources were applied and failed. 

I have created a separate issue to see if we can improve the UX here: https://github.com/GoogleContainerTools/kpt/issues/1964. I think we can close this issue then. The UX might be a bit confusing for people not familiar with the kubectl tool, but it does work as intended.

**Comment by frankfarzan** [comment]
This is both a UX issue and documentation issue. For example, no where in our documentation do we explain any of this:

https://kpt.dev/reference/live/preview/

It's not really acceptable to say that's how kubectl works, and users should already be familiar with it. If that's the case, why are we providing this at all instead of directing users to use kubectl? kubectl is the floor, not the ceiling for the UX we need to provide.

There's also the claim of "improved preview experience" which needs to be shown in practice.
